### PR TITLE
Add support for aws deployment

### LINKED ISSUE #3: Add support AWS
### Describe the improvement request

Currently, the Pulumi codebase for the ephemeral runner only creates resources in GCP.
Modify the repo so that AWS is also supported.
Think about the design of the repo structure and the codebase, so that it can be easily extendable for future cloud providers. This is important because the repo is used either in a custom Github Action (coming in a later ticket) or used directly in a workflow that checks out the runner repo, and executes Pulumi. Pulumi needs to be aware of its context, so it needs to be aware of which cloud provider it will use, and which machine image (machine images will determine which arch is used).

There is a potential threat here, and that is a miss-match between an instance type requested from Pulumi, and a machine image provided. 

So far, needed identified resources for AWS are:
Launch configuration
Autoscaling group
Security group (for debugging purposes)
IAM ??? 
Network resources ??? 

Simple example: https://www.pulumi.com/registry/packages/aws/how-to-guides/ec2-webserver/

### PR DESCRIPTION
**Pull request by obasekiosa**
fixes #3 

This PR Adds Support for the deployment of runners to the AWS cloud,

### CODE REVIEW

File: aws/index.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+import * as awsx from "@pulumi/awsx";
+
+import { FleetConfigureation} from "./instance-template";
```
Comment (pavlovic-ivan) [inline aws/index.ts:5]: Please check the small typo in the the name exported object namel\

File: aws/instance-template.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+import * as fs from "fs";
+
+const LauchConfiguration = new aws.ec2.LaunchTemplate("foo", {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:5]: I would give this launch template a meaningful instead of "foo"

File: aws/instance.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+
+export const FleetConfigureation = new aws.ec2.Fleet("example", {
+    launchTemplateConfig: {
+        launchTemplateSpecification: {
+            launchTemplateId: aws_launch_template.example.id,
```
Comment (pavlovic-ivan) [inline aws/instance.ts:7]: Is this correct? Shouldn't you import the Launch Configuration from file above, and use it's id, and latest version in the line below?

File: aws/instance.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+
+export const FleetConfigureation = new aws.ec2.Fleet("example", {
```
Comment (pavlovic-ivan) [inline aws/instance.ts:4]: I would give this a meaningful name instead of `example`

File: aws/network.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+
+
+export const group = new aws.ec2.SecurityGroup("sec-group-name", {
```
Comment (pavlovic-ivan) [inline aws/network.ts:5]: I would give this a meaningful name as well

File: aws/network.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    ingress: [
+        {protocol: "tcp", fromPort: 22, toPort: 22, cidrBlocks: ["0.0.0.0/0"] },
+    ]
+});
+
+export const subnet = new aws.ec2.Subnet("subnet-name", {
```
Comment (pavlovic-ivan) [inline aws/network.ts:11]: Meaningful name here as well

File: aws/instance-template.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    },
+    ramDiskId: "test",
+    vpcSecurityGroupIds: ["sg-12345678"],
+    tagSpecifications: [{
+        resourceType: "instance",
+        tags: {
+            Name: "test",
+        },
+    }],
+    userData: Buffer.from(fs.readFileSync(`${path.module}/example.sh`), 'binary').toString('base64'),
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:67]: Where is this script `example.sh`?

File: aws/instance-template.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        enabled: true,
+    },
+    networkInterfaces: [{
+        associatePublicIpAddress: "true",
+    }],
+    placement: {
+        availabilityZone: "us-west-2a",
+    },
+    ramDiskId: "test",
+    vpcSecurityGroupIds: ["sg-12345678"],
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:60]: security group should be imported and referred to here

File: aws/instance-template.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    tagSpecifications: [{
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:61]: If tags are going to be set, they should also be meaningful

File: aws/instance-template.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+import * as fs from "fs";
+
+const LauchConfiguration = new aws.ec2.LaunchTemplate("foo", {
+    blockDeviceMappings: [{
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:6]: I think we will not need blcokDeviceMappings. There will be default devices attached by the AMI

File: aws/instance-template.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        deviceName: "/dev/sda1",
+        ebs: {
+            volumeSize: 20,
+        },
+    }],
+    capacityReservationSpecification: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:12]: Not needed

File: aws/instance-template.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        capacityReservationPreference: "open",
+    },
+    cpuOptions: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:15]: This should come with the machine type we choose, so i would remove this.

File: aws/instance-template.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        coreCount: 4,
+        threadsPerCore: 2,
+    },
+    creditSpecification: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:19]: Not needed

File: aws/instance-template.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        cpuCredits: "standard",
+    },
+    disableApiTermination: true,
+    ebsOptimized: "true",
+    elasticGpuSpecifications: [{
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:24]: Not needed

File: aws/instance-template.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        type: "test",
+    }],
+    elasticInferenceAccelerator: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:27]: Not needed

File: aws/instance-template.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        type: "eia1.medium",
+    },
+    iamInstanceProfile: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:30]: Instance profile should be created in the stack and then referred to here

File: aws/instance-template.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        name: "test",
+    },
+    imageId: "ami-test",
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:33]: This should be a value coming from the config file

File: aws/instance-template.ts [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    instanceInitiatedShutdownBehavior: "terminate",
+    instanceMarketOptions: {
+        marketType: "spot",
+    },
+    instanceType: "t2.micro",
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:38]: Instance type should come from the config file

File: aws/instance-template.ts [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    kernelId: "test",
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:39]: Not needed

File: aws/instance-template.ts [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    keyName: "test",
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:40]: Not needed

File: aws/instance-template.ts [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    licenseSpecifications: [{
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:41]: Not needed

File: aws/instance-template.ts [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    monitoring: {
+        enabled: true,
+    },
+    networkInterfaces: [{
+        associatePublicIpAddress: "true",
+    }],
+    placement: {
+        availabilityZone: "us-west-2a",
+    },
+    ramDiskId: "test",
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:59]: not needed

File: aws/instance-template.ts [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+    licenseSpecifications: [{
+        licenseConfigurationArn: "arn:aws:license-manager:eu-west-1:123456789012:license-configuration:lic-0123456789abcdef0123456789abcdef",
+    }],
+    metadataOptions: {
+        httpEndpoint: "enabled",
+        httpTokens: "required",
+        httpPutResponseHopLimit: 1,
+        instanceMetadataTags: "enabled",
+    },
+    monitoring: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:50]: not needed

File: aws/instance-template.ts [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    instanceMarketOptions: {
+        marketType: "spot",
+    },
+    instanceType: "t2.micro",
+    kernelId: "test",
+    keyName: "test",
+    licenseSpecifications: [{
+        licenseConfigurationArn: "arn:aws:license-manager:eu-west-1:123456789012:license-configuration:lic-0123456789abcdef0123456789abcdef",
+    }],
+    metadataOptions: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:44]: not needed

File: aws/instance-template.ts [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+        httpEndpoint: "enabled",
+        httpTokens: "required",
+        httpPutResponseHopLimit: 1,
+        instanceMetadataTags: "enabled",
+    },
+    monitoring: {
+        enabled: true,
+    },
+    networkInterfaces: [{
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:53]: not needed

File: aws/instance-template.ts [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+        associatePublicIpAddress: "true",
+    }],
+    placement: {
```
Comment (pavlovic-ivan) [inline aws/instance-template.ts:56]: not needed

File: aws/instance.ts (hunk 3 above)
Comment (obasekiosa) [reply aws/instance.ts]: oh it was just a placeholder until I got the template working

File: aws/instance-template.ts (hunk 10 above)
Comment (obasekiosa) [reply aws/instance-template.ts:9]: Yes it seems so

File: aws/instance-template.ts (hunk 22 above)
Comment (obasekiosa) [reply aws/instance-template.ts]: I think the availability zone is needed.
The template won't be valid without it

**Pull request merged**